File: Java/animalgame-java/src/Application/Simulation.java
package Application;

import Habitat.*;

public class Simulation {
    private final Habitat habitat;

    public Simulation(Habitat habitat) {
        this.habitat = habitat;
    }

    public void simulate() {
        int rowCount = habitat.getRowCount();
        int columnCount = habitat.getColumnCount();
        Living currentLiving= null;
        Living nextLiving=null;
int row = 0;
int column = 0;
        while(habitat.getLivingCounter()>1)
        {
            if(row > rowCount || column > columnCount)
            {
                throw new IllegalArgumentException("Row and column must be less than the size of the habitat");
            }
            if(currentLiving == null)
            {
                currentLiving = habitat.getLiving(row,column);

            }
            nextLiving = habitat.getLiving(row,column+1);

            if (currentLiving != null && nextLiving != null)
            {
                WarLivings(currentLiving, nextLiving);
                if(habitat.getLiving(currentLiving.getRowLocation(),currentLiving.getColumnLocation()) == null)
                {
                    currentLiving = nextLiving;
                }
                column++;
            }
            if(column >=columnCount-1)
            {
                column = -1;
                row++;
            }
            if(row > rowCount)
            {
                throw new IllegalArgumentException("Row and column must be less than the size of the habitat");
            }

        }

    }

    /***
 *
 * Bitki → Pire
 * Böcek → Bitki
 * Sinek → Pire
 * Bitki → Sinek
 * Sinek → Böcek
 * Böcek →Pire
     *
     * Bitki >> B
     * Sinek >> S
     * Pire  >> P
     * Bocek >> C
     *
     * */
private void WarLivings(Living firstLiving, Living secondLiving) {
    System.out.println("-----------------------");

    System.out.println(firstLiving.Appearance() + " vs " + secondLiving.Appearance());
    // Plant  eats --> Flea
    if (firstLiving.Appearance().equals("B") && secondLiving.Appearance().equals("P")) {
        System.out.println("Plant eats Flea");
        habitat.killLiving(secondLiving.getRowLocation(), secondLiving.getColumnLocation());
    }
    if(firstLiving.Appearance().equals("P") && secondLiving.Appearance().equals("B"))
    {
        System.out.println("Plant eats Flea");
        habitat.killLiving(firstLiving.getRowLocation(), firstLiving.getColumnLocation());
    }
    // Insect  eats --> Plant
    if (firstLiving.Appearance().equals("C") && secondLiving.Appearance().equals("B")) {
        System.out.println("Insect eats Plant");
        habitat.killLiving(secondLiving.getRowLocation(), secondLiving.getColumnLocation());
    }
    if(firstLiving.Appearance().equals("B") && secondLiving.Appearance().equals("C"))
    {
        System.out.println("Insect eats Plant");
        habitat.killLiving(firstLiving.getRowLocation(), firstLiving.getColumnLocation());
    }
    // Mosquito  eats --> Flea
    if (firstLiving.Appearance().equals("S") && secondLiving.Appearance().equals("P")) {
        System.out.println("Mosquito eats Flea");
        habitat.killLiving(secondLiving.getRowLocation(), secondLiving.getColumnLocation());
    }
    if(firstLiving.Appearance().equals("P") && secondLiving.Appearance().equals("S"))
    {
        System.out.println("Mosquito eats Flea");
        habitat.killLiving(firstLiving.getRowLocation(), firstLiving.getColumnLocation());
    }
    // Plant  eats --> Mosquito
    if (firstLiving.Appearance().equals("B") && secondLiving.Appearance().equals("S")) {
        System.out.println("Plant eats Mosquito");
        habitat.killLiving(secondLiving.getRowLocation(), secondLiving.getColumnLocation());
    }
    if(firstLiving.Appearance().equals("S") && secondLiving.Appearance().equals("B"))
    {
        System.out.println("Plant eats Mosquito");
        habitat.killLiving(firstLiving.getRowLocation(), firstLiving.getColumnLocation());
    }
    // Mosquito  eats --> Insect
    if (firstLiving.Appearance().equals("S") && secondLiving.Appearance().equals("C")) {
        System.out.println("Mosquito eats Insect");
        habitat.killLiving(secondLiving.getRowLocation(), secondLiving.getColumnLocation());
    }
    if(firstLiving.Appearance().equals("C") && secondLiving.Appearance().equals("S"))
    {
        System.out.println("Mosquito eats Insect");
        habitat.killLiving(firstLiving.getRowLocation(), firstLiving.getColumnLocation());
    }
    // Insect  eats --> Flea
    if (firstLiving.Appearance().equals("C") && secondLiving.Appearance().equals("P")) {
        System.out.println("Insect eats Flea");
        habitat.killLiving(secondLiving.getRowLocation(), secondLiving.getColumnLocation());
    }
    if(firstLiving.Appearance().equals("P") && secondLiving.Appearance().equals("C"))
    {
        System.out.println("Insect eats Flea");
        habitat.killLiving(firstLiving.getRowLocation(), firstLiving.getColumnLocation());
    }
    if (firstLiving.Appearance().equals(secondLiving.Appearance())) {
        Living loser = compareSameTypeLivings(firstLiving, secondLiving);
        System.out.println("Same type livings are fighting. Loser is: " + loser.Appearance() + " at " + loser.getRowLocation() + "," + loser.getColumnLocation());
        habitat.killLiving(loser.getRowLocation(), loser.getColumnLocation());
    }
    habitat.printHabitat();
    System.out.println("-----------------------");
}
    private Living compareSameTypeLivings(Living living1, Living living2) {
       if(living1.getValue() > living2.getValue())
       {
           return living2;
       }
       else if(living1.getValue() < living2.getValue())
       {
           return living1;
       }
       else
       {
           if(living1.getRowLocation() > living2.getRowLocation())
           {
               return living1;
           }
           else if(living1.getRowLocation() < living2.getRowLocation())
           {
               return living2;
           }
           else
           {
               if(living1.getColumnLocation() > living2.getColumnLocation())
               {
                   return living1;
               }
               else
               {
                   return living2;
               }
           }
       }
    }
}

File: C/src/Simulation.c
#include "Simulation.h"

Simulation newSimulation(Habitat habitat)
{
    Simulation this = (Simulation)malloc(sizeof(struct SIMULATION));
    this->habitat = habitat;
    this->runSimulation = &runSimulation;
    this->deleteSimulation = &deleteSimulation;
    return this;
}
void CompareSameTypeLivings(const Simulation this,Living currentLiving, Living nextLiving)
{
    if(currentLiving->value > nextLiving->value)
    {
        printf("The current living is stronger than the next living");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    else if(currentLiving->value < nextLiving->value)
    {
        printf("The next living is stronger than the current living");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    else
    {
        if(currentLiving->rowLocation > nextLiving->rowLocation)
        {
            this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
        }
        else if(currentLiving->rowLocation < nextLiving->rowLocation)
        {
            this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
        }
        else
        {
            if(currentLiving->columnLocation > nextLiving->columnLocation)
            {
                this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
            }
            else if(currentLiving->columnLocation < nextLiving->columnLocation)
            {
                this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
            }
        }
    
    }
}
    /***
 *
 * Bitki → Pire
 * Böcek → Bitki
 * Sinek → Pire
 * Bitki → Sinek
 * Sinek → Böcek
 * Böcek →Pire
     *
     * Bitki >> B
     * Sinek >> S
     * Pire  >> P
     * Bocek >> C
     *
     * */
void WarLivings(const Simulation this, Living currentLiving, Living nextLiving) {
    printf("--------------------------------- \n ");
    // Plant eats --> Flea
    if (strcmp(currentLiving->appreance(), "B") == 0 && strcmp(nextLiving->appreance(), "P") == 0) {
        printf("Plant eats Flea");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    if (strcmp(currentLiving->appreance(), "P") == 0 && strcmp(nextLiving->appreance(), "B") == 0) {
        printf("Plant eats Flea");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    // Insect eats --> Plant
    if (strcmp(currentLiving->appreance(), "C") == 0 && strcmp(nextLiving->appreance(), "B") == 0) {
        printf("Insect eats Plant");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    if (strcmp(currentLiving->appreance(), "B") == 0 && strcmp(nextLiving->appreance(), "C") == 0) {
        printf("Insect eats Plant");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    // Mosquito eats --> Flea
    if (strcmp(currentLiving->appreance(), "S") == 0 && strcmp(nextLiving->appreance(), "P") == 0) {
        printf("Mosquito eats Flea");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    if (strcmp(currentLiving->appreance(), "P") == 0 && strcmp(nextLiving->appreance(), "S") == 0) {
        printf("Mosquito eats Flea");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    // Plant eats --> Mosquito
    if (strcmp(currentLiving->appreance(), "B") == 0 && strcmp(nextLiving->appreance(), "S") == 0) {
        printf("Plant eats Mosquito");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    if (strcmp(currentLiving->appreance(), "S") == 0 && strcmp(nextLiving->appreance(), "B") == 0) {
        printf("Plant eats Mosquito");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    // Mosquito eats --> Insect
    if (strcmp(currentLiving->appreance(), "S") == 0 && strcmp(nextLiving->appreance(), "C") == 0) {
        printf("Mosquito eats Insect");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    if (strcmp(currentLiving->appreance(), "C") == 0 && strcmp(nextLiving->appreance(), "S") == 0) {
        printf("Mosquito eats Insect");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    // Insect eats --> Flea
    if (strcmp(currentLiving->appreance(), "C") == 0 && strcmp(nextLiving->appreance(), "P") == 0) {
        printf("Insect eats Flea");
        this->habitat->killLiving(this->habitat, nextLiving->rowLocation, nextLiving->columnLocation);
    }
    if (strcmp(currentLiving->appreance(), "P") == 0 && strcmp(nextLiving->appreance(), "C") == 0) {
        printf("Insect eats Flea");
        this->habitat->killLiving(this->habitat, currentLiving->rowLocation, currentLiving->columnLocation);
    }
    // Current living and next living are the same type
    if (strcmp(currentLiving->appreance(), nextLiving->appreance()) == 0) {
        CompareSameTypeLivings(this, currentLiving, nextLiving);
    }
    printf("\n");
   this->habitat->printHabitat(this->habitat);
}
void runSimulation(const Simulation this)
{
    int rowCount = this->habitat->rows;
    int columnCount = this->habitat->cols;
    Living currentLiving= NULL;
    Living nextLiving=NULL;
    int row = 0;
    int column = 0;
        while(this->habitat->livingCount > 1)
        {
            if(row > rowCount || column > columnCount)
            {
                printf("Row and column must be less than the size of the habitat");
            }
            if(currentLiving == NULL)
            {
                currentLiving = this->habitat->habitatMatrix[row][column];

            }
            nextLiving = this->habitat->habitatMatrix[row][column+1];

            if (currentLiving != NULL && nextLiving != NULL)
            {
                WarLivings(this,currentLiving, nextLiving);

                if(this->habitat->habitatMatrix[currentLiving->rowLocation][currentLiving->columnLocation] == NULL)
                {
                    currentLiving = nextLiving;
                }
                column++;
            }
            if(column >=columnCount-1)
            {
                column = -1;
                row++;
            }
            if(row > rowCount)
            {
               printf("Row and column must be less than the size of the habitat");
            }
        }
}
void deleteSimulation(const Simulation this)
{
    free(this);
}
